packages/back-end/src/services/features.ts

```diff
@@ -44,6 +47,55 @@ function generatePayload(
   return defs;
 }
 
+// TODO Put this in a better place?
+function generateVisualExperimentsPayload(
+  visualExperiments: Array<{
+    experiment: ExperimentInterface;
+    visualChangeset: VisualChangesetInterface;
+  }>,
+  _environment: string,
+  _groupMap: GroupMap
+): SDKExperiment[] {
+  return visualExperiments.map(({ experiment: e, visualChangeset: v }) => {
+    const phase = e.phases.slice(-1)[0];
+    const forcedVariation =
+      e.status === "stopped" && e.releasedVariationId
+        ? e.variations.find((v) => v.id === e.releasedVariationId)
+        : null;
+
+    return {
+      key: e.trackingKey,
+      variations: v.visualChanges.map((vc) => ({
+        css: vc.css,
+        domMutations: vc.domMutations,
+      })),
+      hashVersion: 2,
+      hashAttribute: e.hashAttribute,
+      urlPatterns: [v.urlPattern],
+      weights: phase.variationWeights,
+      meta: e.variations.map((v) => ({ key: v.key, name: v.name })),
+      filters: phase.namespace.enabled
+        ? [
+            {
+              attribute: e.hashAttribute,
+              seed: phase.namespace.name,
+              hashVersion: 2,
+              ranges: [phase.namespace.range],
+            },
+          ]
+        : [],
+      seed: phase.seed,
+      name: e.name,
+      phase: `${e.phases.length - 1}`,
+      force: forcedVariation
+        ? e.variations.indexOf(forcedVariation)
+        : undefined,
+      condition: phase.condition,
```

Comment:
This is stored as a string in Mongo, but should be an object here.  We also need to replace saved group ids, just like conditions for features.

Look at `getFeatureDefinition` as an example.

```ts
rule.condition = JSON.parse(
  replaceSavedGroupsInCondition(r.condition, groupMap)
)
```

Reply:
Updated